README.md
---
layout: module
permalink: /module/FreshservicePS/
---
# [FreshservicePS](https://www.flycastpartners.com/FreshservicePS)

<!-- [![PowerShell Gallery](https://img.shields.io/powershellgallery/dt/FreshservicePS.svg?style=for-the-badge)](https://www.powershellgallery.com/packages/FreshservicePS) -->
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

## About
From our continued success as a Platinum-level partner, [Flycast Partners] is considered Freshworks’ preferred partner for complex Freshservice ITSM implementations/integrations as well as a leading source of Freshservice knowledge and services. As part of our commitment to delivering high-quality services, our team of ITIL- and Freshworks-certified ITSM Consultants has created FreshservicePS, a comprehensive Freshservice PowerShell API Library, for diverse operational needs.  

Built by the team who garnered Freshworks Global ITSM Strategic Partner of the Year, Flycast Partners believes in contributing to the Freshworks community. Our PowerShell Modules provide organizations with a versatile solution to administer, build, and streamline their Freshservice environment. By offering our expertise and experience to fellow users, we hope to foster a culture of collaboration and knowledge-sharing that empowers this community to succeed and thrive!

With our FreshservicePS PowerShell Modules, Freshservice Admins have the ability to:

- Make the Freshservice API simple to use with a command-line interface
- Manage Freshservice Ticket, Problems, Changes, Releases, Requesters, and more using simple commands
- Provide additional reporting and filtering capabilities
- Automate re-occurring tasks in Freshservice tenants
- Move data between Freshservice tenants
- Support DevOps and build, test, and deploy solutions in CI/CD operations

The FreshservicePS module aligns closely with the [Freshservice API] documentation, so be sure to reference notes, throttling and other import aspects of the Freshservice API.

## Status - Work in progress

> This project is a **work in progress** and may change significantly before reaching stability based on feedback from the community.
> **Please do not base critical processes on this project** until it has been further refined.  Links to other content may not be working
just yet as we are building content to support this project.

Join the conversation on [FreshservicePS.slack.com](https://join.slack.com/t/freshserviceps/shared_invite/zt-2514m3vky-Azc3DAqh9RhpjJ0ibPwqsQ) -->

---

## Instructions

### Installation

Install FreshservicePS from the [PowerShell Gallery]! `Install-Module` requires PowerShellGet.

```powershell
# One time only install:
Install-Module -Name FreshservicePS -Scope CurrentUser

# Check for updates occasionally:
Update-Module -Name FreshservicePS
```

### Usage

Create a connection profile for FreshservicePS.

```powershell
# To create a configuration to Freshservice:

# Import the module
Import-Module -Name FreshservicePS

# Create a configuration to connect to the Production instance (i.e. https://acme-corp.freshservice.com)
# API and set as the Default connection (automatically connect when importing the module):

New-FreshServiceConnection -Name acme_prod -ApiKey 'gsfdgjkhdfs73jdsbd' -Tenant 'acme-corp' -Environment Production -Default $true

# Create a configuration to connect to the Sandbox (i.e. https://its-fine-fs-sandbox.freshservice.com) 
# API to switch connections using Connect-Freshservice:

New-FreshServiceConnection -Name acme_sbx -ApiKey 'gsfdgjkhdfs73jdsbd' -Tenant 'acme-corp' -Environment Sandbox -Default $false
```

Connect to an environment and execute commands:

```powershell
# When importing the module, it will automatically connect to the default instance 
# (defined with the -Default switch with 'acme_prod'):
Import-Module FreshservicePS

# To switch to another connection, use the Connect-Freshservice cmdlet:
Connect-Freshservice -Name acme_sbx 
```

You can find the full documentation [here](https://flycastpartners.com/FreshservicePS) and in the console.

```powershell
#Review the help at any time!
Get-Help about_FreshservicePS
Get-Command -Module FreshservicePS
Get-Help Get-FreshServiceTicket -Full # or any other command
```

### Rate Limit (Throttling)

Highly recommend reviewing the [rate limit](https://api.freshservice.com/#rate_limit) documentation for Freshservice.  The rate limit applies at the **account level** (not per API Key\account), so all accounts leveraging the API consume calls.  This module performs pagination automatically and pipeline operations that can make many API calls very quickly, so throttling at the module level is default behavior to reduce the probability of a 429 Retry-After which makes the API inaccessible for the entire account for time period (usually 30-40 seconds).  If bulk operations are being performed and no throttling should occur, use the NoThrottling switch:

```powershell
Connect-Freshservice -Name my_instance -NoThrottling
```

For more information on how to use FreshservicePS, tips, blogs and more, check out (COMING SOON!!).

<!-- ### Contribute

Want to contribute to FreshservicePS? Great!
We appreciate who invests their time to make our modules the best they can be. -->

<!-- Check out our guidelines on [Contributing] to our modules and documentation. -->

## Useful links

- [Flycast Partners]
- [Freshservice API]
- [Source Code]
- [Latest Release]
- [Submit an Issue]

- How you can help us: [List of Issues](https://github.com/flycastpartnersinc/FreshservicePS/issues?q=is%3Aissue+is%3Aopen+label%3Aup-for-grabs)

## Disclaimer

Hopefully this is obvious, but:

> This is an open source project (under the [MIT license]), and all contributors are volunteers. All commands are executed at your own risk. **Please ensure have good backups before you start, because you can delete a lot of stuff if you're not careful.**

<!-- reference-style links -->
  [Flycast Partners]: https://www.flycastpartners.com/
  [Freshservice API]: https://api.freshservice.com/
  [PowerShell Gallery]: https://www.powershellgallery.com/
  [Source Code]: https://github.com/flycastpartnersinc/FreshservicePS
  [Latest Release]: https://github.com/flycastpartnersinc/FreshservicePS/releases/latest
  [Submit an Issue]: https://github.com/flycastpartnersinc/FreshservicePS/issues/new
  [MIT license]: https://github.com/flycastpartnersinc/FreshservicePS/blob/main/LICENSE


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.1.6]

### Fixed

- The $IsWindows variable isn't available in older PS versions.  The $FreshServiceConfigPath was using the incorrect variable path.  Updated to check for $env:OS to support pre-core Powershell versions.

## [0.1.5]

### Fixed

- Fixed bug in FreshservicePS.psm1 import not correctly importing functions.

## [0.1.4]

### Added

- Additional support for Workspaces.
- Get-FreshServiceWorkspace cmdlet to list Workspaces.
- Filtering for workspace_id:

  - Get-FreshServiceAgent
  - Get-FreshServiceAsset
  - Get-FreshServiceTicket
  - Get-FreshServiceChange
  - Get-FreshServiceProblem
  - Get-FreshServiceRelease
  - Get-FreshServiceBusinessHour
  - Get-FreshServiceSolutionCategory
  - Get-FreshServiceCatalogCategory
  - Get-FreshServiceCatalogItem
  - Get-FreshServiceCustomObject

- Move methods and options for Workspaces:

  - Set-FreshserviceTicket
  - Set-FreshserviceProblem
  - Set-FreshserviceChange
  - Set-FreshserviceRelease
  - Set-FreshServiceAsset

- Create items in Workspaces:

  - New-FreshServiceAsset
  - New-FreshServiceTicket
  - New-FreshServiceChange
  - New-FreshServiceProblem
  - New-FreshServiceRelease

- Agent licensing for Workspace assignment and License Type:

  - New-FreshServiceAgent
  - Set-FreshServiceAgent

- Updated the New-FreshserviceConnection to only require tenant and apikey.  If a friendly name is not provided, it will use the tenant name.  The first connection will be set to the default.  These settings are to make connection simple for new users.  Converted the Default parameter from Boolean to Switch.

- Originally Freshservice did not support custom statuses and there was a validation done in the cmdlet (e.g. ValidateSet).  Removed the validation to allow custom statuses for tickets.  Current valid values can be viewed with Get-FSTicket -Fields.

- Updated Invoke-FreshserviceRestMethod to append UTF-8 header and set 'application/json; charset=utf-8' and default ContentType.

- Updated Invoke-FreshserviceRestMethod to include -UseBasicParsing for Invoke-WebRequest for all calls to support backwards compatibility for Server Core and older PS versions.

### Fixed

- Filter pagination loop logic.  There is no relative link to perform pagination, updated logic to manually increment page until there is no content returned.

  - Get-FreshServiceAgent
  - Get-FreshServiceAsset
  - Get-FreshserviceTicket
  - Get-FreshServiceProjectTask
  - Get-FreshServiceRequester

## [0.1.3]

### Added

- Removed mandatory parameter for Environment in New-FreshserviceConnection and defaulted to Production

### Fixed

- Moved the if statement for should process to ensure the entire url is built for WhatIf for Remove-FreshserviceAsset

## [0.1.2]

### Fixed

- Updated cmdlets custom_fields with data type object[] to object

## [0.1.1]

### Fixed

- Invoke-FreshServiceRestMethod updates to handle 429 retry and automatically sleep when rate limit is reached.
- Updated incomplete examples help in New-FreshServiceOnboardingRequest and Set-FreshServiceOnboardingRequest

### Added

- Invoke-FreshserviceRestMethod updated to have artificial throttling to ensure module commands do not consume all api credits.  Progressively adds a Sleep after calls from 5 seconds @ 70% consumed to 15 seconds @ 80% consumed to 30 seconds at 90% consumed to avoid 429.
- Connect-Freshservice added a -NoThrottle to override the artificial throttling.

## [0.1.0] 5/24/2023

- Initial release


FreshservicePS/Public/Get-FreshServiceNote.ps1
<#
.SYNOPSIS
    Returns a Freshservice Note on Problem, Change or Release.

.DESCRIPTION
    Returns a Freshservice Note on Problem, Change or Release via REST API. For Tickets,
    see Get-FreshServiceConversation.

    https://api.freshservice.com/#view_a_problem_note
    https://api.freshservice.com/#view_a_change_note
    https://api.freshservice.com/#view_a_release

.PARAMETER Id
    Unique id of the note.

.PARAMETER parent_id
    Unique id of the parent Ticket, Problem, Change or Release.

.PARAMETER type
    Parent object type to create time entry (e.g. Ticket, Problem, Change, Release).

.PARAMETER per_page
    Number of records to return per page during pagination.  Maximum of 100 records.

.PARAMETER page
    The page number to retrieve during pagination.

.EXAMPLE
    Get-FreshServiceNote -parent_id 2 -type Problem

    id            : 21000071787
    created_at    : 1/10/2023 6:24:23 PM
    updated_at    : 1/10/2023 6:24:23 PM
    body          : <div>A note added to a Problem.</div>
    body_text     : A note added to a Problem.
    user_id       : 21000418005
    notify_emails :

    Return all Freshservice Notes on Problem id 2.

.EXAMPLE
    Get-FreshServiceNote -parent_id 2 -type Problem -id 21000071797

    id            : 21000071797
    created_at    : 1/10/2023 7:52:26 PM
    updated_at    : 1/10/2023 8:01:18 PM
    body          : Updated note on a problem
    body_text     : Updated note on a problem
    user_id       : 21000418005
    notify_emails :

    Return a specific Freshservice Note on Problem id 2 by id 21000071797.

.NOTES
    This module was developed and tested with Freshservice REST API v2.
#>
function Get-FreshServiceNote {
    [CmdletBinding(DefaultParameterSetName = 'default')]
    param (
        [Parameter(
            Mandatory = $false,
            HelpMessage = 'Unique id for Ticket to get conversations',
            ParameterSetName = 'default',
            Position = 0
        )]
        [long]$Id,
        [Parameter(
            Mandatory = $true,
            HelpMessage = 'Unique id of the parent Ticket, Problem, Change or Release.',
            ValueFromPipelineByPropertyName = $true
        )]
        [Alias('ParentId')]
        [long]$parent_id,
        [Parameter(
            Mandatory = $true,
            HelpMessage = 'Parent object type to create time entry (e.g. Ticket, Problem, Change, Release)',
            ValueFromPipelineByPropertyName = $true
        )]
        [ValidateSet('Ticket','Problem','Change','Release')]
        [string]$type,
        [Parameter(
            Mandatory = $false,
            HelpMessage = 'Number of records per page returned during pagination.  Default is 30. Max is 100.',
            ParameterSetName = 'default',
            Position = 1
        )]
        [int]$per_page = 100,
        [Parameter(
            Mandatory = $false,
            HelpMessage = 'Page number to begin record return.',
            ParameterSetName = 'default',
            Position = 2
        )]
        [int]$page = 1
    )
    begin {

        $PrivateData  = $MyInvocation.MyCommand.Module.PrivateData

        if (!$PrivateData.FreshserviceBaseUri) {
            throw "No connection found!  Setup a new Freshservice connection with New-FreshServiceConnection and then Connect-FreshService. Set a default connection with New-FreshServiceConnection or Set-FreshConnection to automatically connect when importing the module."
        }

        $qry = [System.Web.HttpUtility]::ParseQueryString([String]::Empty)
        $uri = [System.UriBuilder]('{0}/{1}s/{2}' -f $PrivateData['FreshserviceBaseUri'], $type.ToLower(),$parent_id)
        $enablePagination = $true

    }
    process {

        if ($type -eq 'Ticket') {
            $uri.Path = "{0}/conversations" -f $uri.Path, $Id
        }
        else {
            $uri.Path = "{0}\notes" -f $uri.Path, $Id
        }

        if ($Id) {
            $uri.Path = "{0}/{1}" -f $uri.Path, $Id
            $enablePagination = $true
        }

        try {

            if ($enablePagination) {
                $qry['page'] = $page
                $qry['per_page'] = $per_page
            }

            $uri.Query = $qry.ToString()

            $uri = $uri.Uri.AbsoluteUri

            $results = do {

                $params = @{
                    Uri         = $uri
                    Method      = 'GET'
                    ErrorAction = 'Stop'
                }

                $result = Invoke-FreshworksRestMethod @params

                if ($result.Content) {
                    $content = $result.Content |
                                    ConvertFrom-Json

                    #API returns singluar or plural property based on the number of records, parse to get property returned.
                    $objProperty = $content[0].PSObject.Properties.Name |
                                        Where-Object -FilterScript {$_ -like 'note*' -or $_ -like 'conversation*'}
                    Write-Verbose -Message ("Returning {0} property with count {1}" -f $objProperty, $content."$($objProperty)".Count)
                    $content."$($objProperty)"
                }

                if ($result.Headers.Link) {
                    $uri = [regex]::Matches($result.Headers.Link,'<(?<Uri>.*)>')[0].Groups['Uri'].Value
                }

            }
            until (!$result.Headers.Link)

        }
        catch {
            Throw $_
        }

    }
    end {

        $results

    }
}
